atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/expectations/KeyValueServiceDataTracker.java

```diff
@@ -0,0 +1,129 @@
+import java.util.concurrent.atomic.LongAdder;
+import java.util.function.Function;
+
+public final class KeyValueServiceDataTracker {
+    private final AtomicLongMap<TableReference> bytesReadByTable = AtomicLongMap.create();
+    private final LongAdder bytesReadOverall = new LongAdder();
+    private final AtomicLongMap<TableReference> kvsCallByTable = AtomicLongMap.create();
+    private final LongAdder kvsCallsOverall = new LongAdder();
+    private final ConcurrentMap<TableReference, KvsCallReadInfo> maximumBytesKvsCallInfoByTable =
+            new ConcurrentHashMap<>();
+    private final AtomicReference<Optional<KvsCallReadInfo>> maximumBytesKvsCallInfoOverall =
+            new AtomicReference<>(Optional.empty());
+
+    public TransactionReadInfo getReadInfo() {
+        return ImmutableTransactionReadInfo.builder()
+                .bytesRead(bytesReadOverall.longValue())
+                .kvsCalls(kvsCallsOverall.longValue())
+                .maximumBytesKvsCallInfo(maximumBytesKvsCallInfoOverall.get())
+                .build();
+    }
+
+    /**
+     * This is un-synchronized as it is called after task completion and the abort/commit stage.
+     * For returned statistics, no consistency guarantees are provided if the user interacts with the transaction
+     * outside the context of {@link com.palantir.atlasdb.transaction.api.TransactionTask} or after commit/abort
+     * stages (e.g. this can happen with futures/iterators).
+     */
+    public ImmutableMap<TableReference, TransactionReadInfo> getReadInfoByTable() {
+        Set<TableReference> tableRefs = ImmutableSet.copyOf(Sets.intersection(
+                bytesReadByTable.asMap().keySet(), kvsCallByTable.asMap().keySet()));
+
+        return tableRefs.stream()
+                .collect(ImmutableMap.toImmutableMap(
+                        Function.identity(), tableRef -> ImmutableTransactionReadInfo.builder()
+                                .bytesRead(bytesReadByTable.get(tableRef))
+                                .kvsCalls(kvsCallByTable.get(tableRef))
+                                .maximumBytesKvsCallInfo(
+                                        Optional.ofNullable(maximumBytesKvsCallInfoByTable.get(tableRef)))
+                                .build()));
+    }
+
+    /**
+     * Tracks an effectively completed kvs read method call for some {@link TableReference}.
+     * Effectively completed refers to an eager call (i.e. does not spawn futures or iterators for later consumption).
+     */
+    public void readForTable(TableReference tableRef, String methodName, long bytesRead) {
+        KvsCallReadInfo callInfo = ImmutableKvsCallReadInfo.builder()
+                .bytesRead(bytesRead)
+                .methodName(methodName)
+                .build();
+        updateKvsMethodOverallTallies(callInfo);
+        updateKvsMethodByTableTallies(tableRef, callInfo);
+    }
+
+    /**
+     * Track a lazy kvs read method call for some {@link TableReference}.
```

Comment:
Discussed offline: this isn't actually what this method is for! Lets have a more descriptive javadoc that explains when you expect to call this method